contrib/babelfishpg_tsql/src/pl_comp.c

```diff
@@ -1642,6 +1667,203 @@ pltsql_param_ref(ParseState *pstate, ParamRef *pref)
+		isGeospatialFunction = false;	/* resetting the global variable */
+		return NULL;
+	}
+
+	last_field = (Node *) llast(cref->fields);
+	/* 
+	 * checking if last field matches any of the geospatial functions - stx, sty, lat, long
+	 * and also checking if isGeospatialFunction variable is false to avoid entering the loop if it is a column rather than function
+	 */
+	if (IsA(last_field, String) && !isGeospatialFunction &&
+		(pg_strcasecmp(strVal(last_field), "stx") == 0 ||
+		pg_strcasecmp(strVal(last_field), "sty") == 0 ||
+		pg_strcasecmp(strVal(last_field), "lat") == 0 ||
+		pg_strcasecmp(strVal(last_field), "long") == 0))
+	{
+		FuncExpr   *funcexpr = makeNode(FuncExpr);	/* func expression to modify colref to funcref */
+		char *name = strVal(last_field);
+		Oid funcid;
+		Oid dataTypeId;
+		ColumnRef *colref = makeNode(ColumnRef);	/* new colref to avoid making changes to existing structure */
+
+		isGeospatialFunction = true;
+		colref->fields = list_copy(cref->fields);
+		/* if length is 3 then colref would have been modified by rewrite_column_refs function */
+		if(list_length(colref->fields) == 3)
+		{
+			Node	   *schema = (Node *) linitial(colref->fields);
+			char	   *cur_db = get_cur_db_name();
+			String	   *new_schema;
+
+			if (!is_shared_schema(strVal(schema)))
+			{
+				new_schema = makeString(resolve_schema_name(cur_db, strVal(schema), get_migration_mode()));
+				colref->fields = list_delete_first(colref->fields);
+				colref->fields = lcons(new_schema, colref->fields);
+			}
+		}
+		colref->fields = list_delete_last(colref->fields);	/* deletes the function name and check if the remaining colref is a valid column reference */
+		PG_TRY();
+		{
+			col = transformExpr(pstate, (Node *) colref, pstate->p_expr_kind);	/* Recursive call to check for valid column reference */
+		}
+		PG_CATCH();
+		{
+			isGeospatialFunction = false;	/* resetting the global variable */
+			return NULL;	/* if not a valid column ref */
+		}
+		PG_END_TRY();
+		
+		dataTypeId = ((Var *)col)->vartype;
+		/* Modifying colref to funcref since a valid Geospatial function call is identified */
+		funcid = LookupFuncName(list_make2(makeString("sys"), makeString(name)), 1, &dataTypeId, false);
+		funcexpr->funcid = funcid;
+		funcexpr->funcresulttype = get_func_rettype(funcid);
+		funcexpr->funcretset = false;
+		funcexpr->funcvariadic = false;
+		funcexpr->funcformat = COERCE_EXPLICIT_CALL;
+		funcexpr->args = list_make1(col);
+		funcexpr->location = cref->location;
+		isGeospatialFunction = false;	/* resetting the global variable */
+
+		return (Node *) funcexpr;
+	}
+
+	isGeospatialFunction = false;	/* resetting the global variable */
+
+	return NULL;
+}
+
+/*
+ * This function identifies and modifies geospatial function call for
+ * STDistance(), STAsText(), STAsBinary() functions
+ * TODO: Add handling for more such TSQL function syntax here
+ */
+static List *resolve_geospatial_func_ref(ParseState *pstate, FuncCall *fn, List *fargs)
```

Comment:
It's possible to merge both resolve_geospatial_column and resolve_geospatial_func. You can use the node type to distinguish between column ref and function call